content/dharma/abstract-data-structures.md

```diff
@@ -0,0 +1,118 @@
+
+![node and linked list](./node-linked-list.png)
+
+Linked-list ADT would support the following operations:
+
+```js
+// general operations
+getHead(); // returns back head node
+getSize(); // return the current size of the linked-list
+isEmpty(); // returns true if the linked-list is empty
+
+// insert and replace operations
+insertBeginning(value); // inserts new value at the beginning of the linked-list
+insertEnd(value); // inserts new value at the end of the linked-list
+insertAtPosition(value, index); // inserts new value at the given positional index
+replaceAtPosition(value, index); // replaces the value at give index with the new value
+
+// delete operations
+deleteBeginning(); // removes the first value and returns its value
+deleteEnd(); // removes the last value and returns its value
+deleteAtPosition(index); // removes node from the given positional index and returns its value
+
+// traverse, sort and search operations
+traverse(); // goes through all the elements once in the linked list and prints them
+search(value); // searches given value and returns true if the value is found in linked-list
+sort(order); // sorts the linked-list in the given order (ascending/descending)
+retrieve(index); // returns the value stored at the given index location
+```
+
+## Stack ADT
+
+A **Stack** is a linear data structure, in which data is entered and removed from _only a single point_. This point is called the **top** of the stack. It follows the last-in, first-out (colloquially abbreviated as LIFO) format for storing and discarding data. This means that the last element added to the top of the stack is the first element that will be removed from the stack. There is no other way to access other elements in the stack but the element that is at the top of the stack.
+
+![Stack ADT](./stack-adt.png)
+
+Stack ADT supports the following operations:
+
+```js
+push(value); // inserts a new value at the top of the stack
+pop(); // removes the element stored at the top of the stack and returns it
+peek(); // returns the top element without removing it from the stack
+```
+
+## Queue ADT
+
+A **Queue** is a linear data structure, in which data is inserted from one end and removed from the other end. The place where the data is inserted is called the **rear** end of the queue and the place from which the data is removed is called the **front** of the queue. It follows the first-in, first-out (abbreviated as FIFO) configuration for storing and removing data. This means that the data that entered first, also leaves first out of the queue. There is no other way to access other elements in the queue but the front of the queue.
+
+![Queue ADT](./queue-adt.png)
+
+Queue ADT supports the following operations:
+
+```js
+enqueue(value); // inserts a new value in the queue
+dequeue(); // removes the element at the front of the queue and returns it
+```
```

Comment:
It is a good practice to have a concluding section that does a high level recap of what was gone over in the article.

Reply:
I'll add the concluding section with a high level review of the data types discussed above.